src/module_4/non_linear_modeling.md

```diff
@@ -0,0 +1,1794 @@
+pred_val = ca.predict_proba(X_val)[:, 1]
+```
+
+
+```python
+n_trees_grid = [200, 210]
+ca_train_aucs_list = []
+ca_val_aucs_list = []
+ca_train_aps_list = []
+ca_val_aps_list = []
+
+for lr in [0.05]:
+    for depth in [5, 7]:
+        for n_tree in n_trees_grid:
+            ca = CatBoostClassifier(
+                max_depth=depth, learning_rate=lr, n_estimators=n_tree, verbose=0
+            )
+            ca.fit(X_train, y_train, cat_features)
+            evaluate_model(
+                ca,
+                X_train,
+                y_train,
+                X_val,
+                y_val,
+                ca_train_aucs_list,
+                ca_val_aucs_list,
+                ca_train_aps_list,
+                ca_val_aps_list,
+            )
+
+            print(
+                f"CatBoost n_trees={n_tree}, max_depth = {depth}, learning_rate = {lr} :"
+                f"train_auc= {ca_train_aucs_list[-1]:.4f} and val_auc= {ca_val_aucs_list[-1]:.4f}."
+                f" Train_ap= {ca_train_aps_list[-1]:.4f} and val_ap= {ca_val_aps_list[-1]:.4f}."
+            )
+```
+
+    CatBoost n_trees=200, max_depth = 5, learning_rate = 0.05 :train_auc= 0.8437 and val_auc= 0.8450. Train_ap= 0.2146 and val_ap= 0.2012.
+    CatBoost n_trees=210, max_depth = 5, learning_rate = 0.05 :train_auc= 0.8441 and val_auc= 0.8456. Train_ap= 0.2154 and val_ap= 0.2014.
+    CatBoost n_trees=200, max_depth = 7, learning_rate = 0.05 :train_auc= 0.8476 and val_auc= 0.8479. Train_ap= 0.2301 and val_ap= 0.2020.
+    CatBoost n_trees=210, max_depth = 7, learning_rate = 0.05 :train_auc= 0.8479 and val_auc= 0.8480. Train_ap= 0.2315 and val_ap= 0.2025.
+
+
+# Curves of the best models
+
+As I have already studied the best models taking into account to avoid overfitting, this curves represent the predictions with validation data for the best model of each type
+
+
+```python
+lr = make_pipeline(StandardScaler(), LogisticRegression(penalty="l2", C=1e-6))
+lr.fit(X_train[lr_filtered_cols], y_train)
+lr_predictions = lr.predict_proba(X_val[lr_filtered_cols])[:, 1]
+print("LR Ridge trained with filtered columns")
+
+dt = DecisionTreeClassifier(max_depth=8)
+dt.fit(X_train[dt_filtered_cols], y_train)
+dt_predictions = dt.predict_proba(X_val[dt_filtered_cols])[:, 1]
+print("Decission Tree trained with filtered columns")
+
+rf = RandomForestClassifier(n_estimators=n_tree)
+rf.fit(X_train[train_cols], y_train)
+rf_predictions = rf.predict_proba(X_val[train_cols])[:, 1]
+print("Random Forest trained with all the columns")
+
+xg = XGBClassifier(max_depth=7, learning_rate=0.05, n_estimators=100)
+xg.fit(X_train[train_cols], y_train)
+xg_predictions = xg.predict_proba(X_val[train_cols])[:, 1]
+print("XGBoost trained with all the columns")
+
+ca = CatBoostClassifier(max_depth=7, learning_rate=0.05, n_estimators=210, verbose=0)
+ca.fit(X_train, y_train, cat_features)
+ca_predictions = ca.predict_proba(X_val)[:, 1]
+print("CatBoost trained with all the columns")
```

Comment:
Maybe we could've used a loop to iterate through the classifiers to avoid repetition